File: frontend/apps/ios/Pantopus/Features/Scheduling/Bookings/Extras/Waitlist/WaitlistJoinSheet.swift
//
//  WaitlistJoinSheet.swift
//  Pantopus
//
//  Stream I9 — E13 Waitlist (invitee join). A host-branded bottom sheet shown
//  when an event type / slot is full: a "Fully booked" pill, the requested
//  window, contact fields, and a "Join waitlist" CTA that swaps to a joined
//  confirmation. Presented by the public booking flow (I5/I7) when full.
//  Accent follows the host's pillar.
//

import SwiftUI

struct WaitlistJoinSheet: View {
    @State private var viewModel: WaitlistJoinViewModel
    var accent: Color
    let onClose: () -> Void

    init(viewModel: WaitlistJoinViewModel, accent: Color, onClose: @escaping () -> Void) {
        _viewModel = State(wrappedValue: viewModel)
        self.accent = accent
        self.onClose = onClose
    }

    var body: some View {
        VStack(spacing: 0) {
            ExtrasSheetGrabber()
            if viewModel.didJoin {
                joinedConfirmation
            } else {
                form
                footer
            }
        }
        .background(Theme.Color.appSurface)
        .presentationDetents([.large])
        .presentationDragIndicator(.hidden)
        .accessibilityIdentifier("scheduling.waitlistJoin")
    }

    // MARK: Form

    private var form: some View {
        @Bindable var viewModel = viewModel
        return ScrollView {
            VStack(alignment: .leading, spacing: Spacing.s3) {
                fullyBookedPill
                Text(viewModel.windowLabel)
                    .font(.system(size: 16.5, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                Text("Join the waitlist and we'll email you the moment a spot opens.")
                    .font(.system(size: 12.5))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .lineSpacing(2)

                tzChip

                field(label: "Your name", icon: .user, placeholder: "Full name", text: $viewModel.name)
                field(
                    label: "Email",
                    icon: .mail,
                    placeholder: "For a message when a spot opens",
                    text: $viewModel.email,
                    isEmail: true
                )
                field(label: "Preferred time", icon: nil, placeholder: "Any morning works (optional)", text: $viewModel.preferredTime)

                if let message = viewModel.errorMessage {
                    ExtrasInlineError(message: message)
                }
            }
            .padding(.horizontal, Spacing.s4)
            .padding(.top, Spacing.s2)
            .padding(.bottom, Spacing.s3)
        }
    }

    private var fullyBookedPill: some View {
        HStack(spacing: Spacing.s2 - 2) {
            Icon(.usersRound, size: 13, color: Theme.Color.warning)
            Text("Fully booked")
                .font(.system(size: 11, weight: .bold))
                .foregroundStyle(Theme.Color.warning)
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s1 + 1)
        .background(Theme.Color.warningBg)
        .overlay(Capsule().strokeBorder(Theme.Color.warningLight, lineWidth: 1))
        .clipShape(Capsule())
    }

    private var tzChip: some View {
        HStack(spacing: Spacing.s2 - 2) {
            Icon(.globe, size: 13, color: Theme.Color.appTextStrong)
            Text("Times in \(viewModel.timeZoneLabel)")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(Theme.Color.appTextStrong)
        }
        .padding(.horizontal, Spacing.s3)
        .frame(height: 28)
        .overlay(Capsule().strokeBorder(Theme.Color.appBorder, lineWidth: 1))
        .clipShape(Capsule())
    }

    private func field(
        label: String,
        icon: PantopusIcon?,
        placeholder: String,
        text: Binding<String>,
        isEmail: Bool = false
    ) -> some View {
        VStack(alignment: .leading, spacing: Spacing.s2 - 2) {
            ExtrasOverline(text: label)
            HStack(spacing: Spacing.s2 + 1) {
                if let icon { Icon(icon, size: 16, color: Theme.Color.appTextMuted) }
                TextField(placeholder, text: text)
                    .font(.system(size: 13))
                    .foregroundStyle(Theme.Color.appText)
                    .keyboardType(isEmail ? .emailAddress : .default)
                    .textInputAutocapitalization(isEmail ? .never : .words)
                    .autocorrectionDisabled(isEmail)
            }
            .padding(.horizontal, Spacing.s3)
            .frame(height: 44)
            .background(Theme.Color.appSurfaceSunken)
            .overlay(
                RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
                    .strokeBorder(Theme.Color.appBorder, lineWidth: 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: Radii.md, style: .continuous))
        }
    }

    private var footer: some View {
        ExtrasStickyFooter {
            ExtrasSolidButton(
                title: "Join waitlist",
                icon: .userPlus,
                accent: accent,
                isEnabled: viewModel.canJoin,
                isBusy: viewModel.isJoining
            ) {
                Task { await viewModel.join() }
            }
        }
    }

    // MARK: Joined

    private var joinedConfirmation: some View {
        VStack(spacing: Spacing.s4) {
            Spacer(minLength: Spacing.s8)
            ExtrasIconDisc(
                icon: .check,
                background: Theme.Color.successBg,
                foreground: Theme.Color.success,
                diameter: 74
            )
            VStack(spacing: Spacing.s2) {
                Text("You're on the waitlist")
                    .font(.system(size: 17.5, weight: .bold))
                    .foregroundStyle(Theme.Color.appText)
                Text("We'll email you the moment a spot opens.")
                    .font(.system(size: 13))
                    .foregroundStyle(Theme.Color.appTextSecondary)
                    .multilineTextAlignment(.center)
            }
            Spacer()
            ExtrasGhostButton(title: "Done") { onClose() }
                .padding(.horizontal, Spacing.s4)
                .padding(.bottom, Spacing.s5)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .padding(.horizontal, Spacing.s5)
    }
}

#if DEBUG
#Preview {
    Color.black.sheet(isPresented: .constant(true)) {
        WaitlistJoinSheet(
            viewModel: WaitlistJoinViewModel(
                slug: "acme-studio",
                eventTypeSlug: "group-class",
                windowLabel: "Sat, Jun 14 · 10:00 AM",
                timeZoneLabel: "Pacific Time",
                client: .shared
            ),
            accent: Theme.Color.business
        ) {}
    }
}
#endif

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/scheduling/bookings_extra/WaitlistJoinSheet.kt
@file:Suppress(
    "PackageNaming",
    "LongMethod",
    "LongParameterList",
    "TooManyFunctions",
    "CyclomaticComplexMethod",
    "LargeClass",
    "MatchingDeclarationName",
)
@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package app.pantopus.android.ui.screens.scheduling.bookings_extra

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.SheetState
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import app.pantopus.android.ui.components.GhostButton
import app.pantopus.android.ui.components.PrimaryButton
import app.pantopus.android.ui.theme.PantopusColors
import app.pantopus.android.ui.theme.PantopusIcon
import app.pantopus.android.ui.theme.PantopusIconImage
import app.pantopus.android.ui.theme.PantopusTextStyle
import app.pantopus.android.ui.theme.Radii
import app.pantopus.android.ui.theme.Spacing

const val WAITLIST_JOIN_TAG = "scheduling.waitlistJoin"

/**
 * E13 (invitee) Waitlist Join sheet — a host-branded bottom sheet shown when an
 * event/slot is full. Collects a name + email (+ optional preferred time) and
 * joins the public waitlist; on success it swaps to an A18-style "you're on the
 * waitlist" confirmation. Stateless: the public flow drives it via
 * [WaitlistJoinViewModel].
 */
@Composable
internal fun WaitlistJoinSheet(
    state: WaitlistJoinUiState,
    sheetState: SheetState,
    accent: Color,
    onName: (String) -> Unit,
    onEmail: (String) -> Unit,
    onPreferredTime: (String) -> Unit,
    onJoin: () -> Unit,
    onDismiss: () -> Unit,
) {
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = PantopusColors.appSurface,
        modifier = Modifier.testTag(WAITLIST_JOIN_TAG),
    ) {
        if (state.didJoin) {
            JoinedConfirmation(accent = accent, onDone = onDismiss)
        } else {
            Column(
                modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.s4, vertical = Spacing.s2),
                verticalArrangement = Arrangement.spacedBy(Spacing.s4),
            ) {
                FullyBookedPill()
                Column(verticalArrangement = Arrangement.spacedBy(Spacing.s1)) {
                    Text(
                        text =
                            state.windowLabel.ifBlank {
                                "This time is full"
                            },
                        style = PantopusTextStyle.h3,
                        fontWeight = FontWeight.Bold,
                        color = PantopusColors.appText,
                    )
                    Text(
                        text = "Join the waitlist and we'll text you the moment a spot opens.",
                        style = PantopusTextStyle.small,
                        color = PantopusColors.appTextSecondary,
                    )
                }
                if (state.timezoneLabel.isNotBlank()) {
                    TimezoneChip(label = state.timezoneLabel)
                }

                LabeledField("Your name") {
                    ExtrasInputField(
                        value = state.name,
                        onValueChange = onName,
                        placeholder = "Full name",
                        leadingIcon = PantopusIcon.User,
                        accent = accent,
                    )
                }
                LabeledField("Email") {
                    ExtrasInputField(
                        value = state.email,
                        onValueChange = onEmail,
                        placeholder = "For an alert when a spot opens",
                        leadingIcon = PantopusIcon.Mail,
                        accent = accent,
                        keyboardType = KeyboardType.Email,
                    )
                }
                LabeledField("Preferred time") {
                    ExtrasInputField(
                        value = state.preferredTime,
                        onValueChange = onPreferredTime,
                        placeholder = "Any morning works (optional)",
                        leadingIcon = PantopusIcon.Clock,
                        accent = accent,
                    )
                }

                state.error?.let { ExtrasInlineError(message = it) }

                PrimaryButton(
                    title = "Join waitlist",
                    onClick = onJoin,
                    modifier = Modifier.fillMaxWidth().padding(bottom = Spacing.s4),
                    isLoading = state.joining,
                    isEnabled = state.canJoin,
                )
            }
        }
    }
}

@Composable
private fun LabeledField(
    label: String,
    content: @Composable () -> Unit,
) {
    Column(verticalArrangement = Arrangement.spacedBy(Spacing.s2)) {
        ExtrasOverline(label)
        content()
    }
}

@Composable
private fun FullyBookedPill() {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.warningBg)
                .border(1.dp, PantopusColors.warningLight, RoundedCornerShape(Radii.pill))
                .padding(horizontal = Spacing.s3, vertical = Spacing.s1),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(icon = PantopusIcon.Users, contentDescription = null, size = 13.dp, tint = PantopusColors.warning)
        Text(text = "Fully booked", style = PantopusTextStyle.caption, fontWeight = FontWeight.SemiBold, color = PantopusColors.warning)
    }
}

@Composable
private fun TimezoneChip(label: String) {
    Row(
        modifier =
            Modifier
                .clip(RoundedCornerShape(Radii.pill))
                .background(PantopusColors.appSurfaceSunken)
                .padding(horizontal = Spacing.s3, vertical = Spacing.s2),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Spacing.s1),
    ) {
        PantopusIconImage(icon = PantopusIcon.Globe, contentDescription = null, size = 13.dp, tint = PantopusColors.appTextSecondary)
        Text(text = label, style = PantopusTextStyle.caption, fontWeight = FontWeight.Medium, color = PantopusColors.appTextStrong)
    }
}

@Composable
private fun JoinedConfirmation(
    accent: Color,
    onDone: () -> Unit,
) {
    Column(
        modifier = Modifier.fillMaxWidth().padding(horizontal = Spacing.s4, vertical = Spacing.s6),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Spacing.s3),
    ) {
        Box(
            modifier = Modifier.size(64.dp).clip(CircleShape).background(PantopusColors.successBg),
            contentAlignment = Alignment.Center,
        ) {
            PantopusIconImage(icon = PantopusIcon.Check, contentDescription = null, size = 30.dp, tint = PantopusColors.success)
        }
        Text(
            text = "You're on the waitlist",
            style = PantopusTextStyle.h3,
            fontWeight = FontWeight.Bold,
            color = PantopusColors.appText,
            textAlign = TextAlign.Center,
        )
        Text(
            text = "We'll text you the moment a spot frees up.",
            style = PantopusTextStyle.small,
            color = PantopusColors.appTextSecondary,
            textAlign = TextAlign.Center,
        )
        GhostButton(title = "Done", onClick = onDone, modifier = Modifier.fillMaxWidth().padding(top = Spacing.s2, bottom = Spacing.s4))
    }
}
